### PR TITLE
[Refactor:TAGrading] Events Infrastructure & StatusBanner

### PR DESCRIPTION
**Pull request by jndlansh**
### Why is this Change Important & Necessary?

In this PR we set up the necessary infrastructure required for Vue migration.
- Use props and emits in the Vue component.
- The Vue component should not deal with external DOM directly
- Just emit from the Vue component and our infrastructure will handle this in Component mounting in twig.
- It uses an Events mapping in Twig file :
```
'events': {
    'color-change': '(event) => { document.body.style.background = event; }',
    'position-change': '(event) => { handlePositionChange(event.panelId, event.panelPosition); }',
}

```
**PLUS** 

In the ongoing migration of Submitty's UI from legacy Twig + jQuery to Vue.js, in this PR we migrate the `StatusBanner`  to vue.

### What is the New Behavior?
  ErrorMessage.twig is replaced by StatusBanner.vue, a pure display component that receives message and color as props
  and emits color-change on mount so the parent can set the body background color. The component renders a colored banner
   in the TA grading panel header showing submission status (late, no submission, withdrawn, overridden, etc.). No
  behavioral change, the banner looks and works identically.

https://github.com/user-attachments/assets/3c04f46a-32c4-4431-b744-cfe2e65200ec

### What steps should a reviewer take to reproduce or test the bug or new feature?
1. Navigate to a TA grading page where the student has a submission status condition (late, no submission, withdrawn,
  overridden grades, etc.)
2. Verify the colored banner appears in the grading panel header with the correct message and background color
### Automated Testing & Documentation
- Added Cypress Component Testing.
image

### Other information

- Not a Breaking change
- No Migrations needed.

### CODE REVIEW

File: site/app/templates/Vue.twig [hunk 1]
```diff
@@ -11,6 +11,7 @@
         "#vue-{{ id }}",
         "{{ type }}",
         "{{ name }}",
-        {{ args | default({}) | json_encode() | raw() }}
+        {{ args | default({}) | json_encode() | raw() }},
+        {{ callbacks | default({}) | json_encode() | raw() }}
```
Comment (williamjallen) [inline site/app/templates/Vue.twig]: Let's call this "events" instead of "callbacks".

File: site/vue/src/components/StatusBanner.vue [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+const props = defineProps<{
+    message: string;
+    color: string;
+}>();
+
+onMounted(() => {
+    window.dispatchEvent(new CustomEvent('status-banner-color-change', {
+        detail: props.color,
+    }));
+});
```
Comment (williamjallen) [inline site/vue/src/components/StatusBanner.vue:15]: We shouldn't be emitting window events from Vue components IMO.  The Vue component should emit Vue [events](https://vuejs.org/guide/components/events.html), and the wrapper should handle catching those.

File: site/vue/src/callbacks.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+window.handleStatusBannerColorChange = function (color: string) {
+    document.body.style.background = color;
+};
+
+declare global {
+    interface Window {
+        handleStatusBannerColorChange(color: string): void;
+    }
+}
```
Comment (williamjallen) [inline site/vue/src/callbacks.ts]: I'm confused about what's going on here.

**Review by williamjallen — CHANGES_REQUESTED** [review]

File: site/vue/src/components/StatusBanner.vue (hunk 2 above)
Comment (jndlansh) [reply site/vue/src/components/StatusBanner.vue:15]: Got it, solved in newer approach - now vue component just emits and the business logic is handled in separate file (elsewhere)

File: site/app/templates/Vue.twig (hunk 1 above)
Comment (jndlansh) [reply site/app/templates/Vue.twig]: Alright

File: site/vue/src/callbacks.ts (hunk 3 above)
Comment (jndlansh) [reply site/vue/src/callbacks.ts]: callbacks.ts removed - business logic now handled in pre-existing ta-grading-init.ts file

File: site/cypress/component/StatusBanner.cy.js [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+        });
+
+        it('applies the correct background color via inline style', () => {
+            cy.mount(StatusBanner, { props: defaultProps });
+            cy.get('[data-testid="status-banner"]')
+                .should('have.attr', 'style')
+                .and('contain', 'background-color: var(--standard-medium-orange)');
+        });
+
+        it('sets text color to black', () => {
```
Comment (JManion32) [inline site/cypress/component/StatusBanner.cy.js]: Do we need to test if the color is black if it's hardcoded? Seems unnecessary unless it was dynamic like the background is.

File: site/cypress/component/StatusBanner.cy.js [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+            cy.mount(StatusBanner, { props: defaultProps });
+            cy.get('[data-testid="status-banner"]')
+                .should('have.css', 'color', 'rgb(0, 0, 0)');
+        });
+
+        it('has the bar_banner id for legacy DOM compatibility', () => {
```
Comment (JManion32) [inline site/cypress/component/StatusBanner.cy.js]: I don't think we need to verify the ID exists here. This only protects against a developer removing the `id` from the Vue code. When the refactor is complete it will have no purpose. This also won't scale well. Asserting ALL ids throughout our new Vue components would add hundreds of lines of bloat to the component tests.

File: site/cypress/component/StatusBanner.cy.js [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+                    },
+                });
+                cy.mount(Wrapper);
+                cy.get('@multiMountHandler').should('have.callCount', 1);
+                cy.mount(Wrapper);
+                cy.get('@multiMountHandler').should('have.callCount', 2);
+            });
+        });
+
+        it('does not dispatch a window CustomEvent', () => {
```
Comment (JManion32) [inline site/cypress/component/StatusBanner.cy.js]: I don't think this test belongs here. While I agree that components shouldn't be calling `window.dispatchEvent()`, adding this assertion to an individual component test implies we'd need to duplicate it across every component test file.

Perhaps we could define some helpers that we import to each component test (maybe several helpers, combined into a single import), that check for things like this.

While tests like this may not be useful for us now that we figured out the event infrastructure, I think they could serve as guardrails for future developers who pick up the refactor since an LLM could lead them to the same pitfalls that we encountered.

These tests are low priority right now. Let's just remove it, and put this idea on the backburner.

**Review by JManion32 — CHANGES_REQUESTED** [review]
Nice work updating the description.

Functionality looks great, I only have requested changes (or I guess just removals) for the component tests. `main.ts` looks clean, and the new Vue component is using props and emits rather than relying on Submitty-specific behavior and functions.

A comment about the `camelize` and `capitalize` in `main.ts`:
We talked about this a little bit in our meeting today. Based on a quick search, this implementation with the `camelize` and `capitalize` is correct, since Vue expects event names to be kebab case and automatically maps them to camel case listeners. Looking at some older Vue components in the codebase, it looks like there is some inconsistency with event naming conventions that should probably be cleaned up in a separate PR.

File: site/cypress/component/StatusBanner.cy.js (hunk 6 above)
Comment (jndlansh) [reply site/cypress/component/StatusBanner.cy.js]: yes, this should be removed

File: site/cypress/component/StatusBanner.cy.js (hunk 4 above)
Comment (jndlansh) [reply site/cypress/component/StatusBanner.cy.js]: I can remove that, if you say. But it does not make much of a diff, or does it?

File: site/cypress/component/StatusBanner.cy.js (hunk 5 above)
Comment (jndlansh) [reply site/cypress/component/StatusBanner.cy.js]: Alright, will remove this

File: site/cypress/component/StatusBanner.cy.js (hunk 4 above)
Comment (JManion32) [reply site/cypress/component/StatusBanner.cy.js]: No, it doesn't make much of a difference, but the point of these tests is to see how they render based on their props, and what they emit. This color will never change since it's not tied to any state. We need to keep these files as lean as possible, since the more tests we add, the more code we have to maintain. In the future, if this color were to become tied to state, then the component test should be adjusted accordingly.

Comment (jndlansh) [reply site/cypress/component/StatusBanner.cy.js]: Got it. Removed!

File: site/cypress/component/StatusBanner.cy.js [hunk 7]
```diff
@@ -0,0 +1,144 @@
+import { h, defineComponent } from 'vue';
+import StatusBanner from '../../vue/src/components/StatusBanner.vue';
+
+const defaultProps = {
+    message: 'Late Submission (Using one or more late days)',
+    color: 'var(--standard-medium-orange)',
+};
+
+function mountWithEmitSpy(props) {
```
Comment (JManion32) [inline site/cypress/component/StatusBanner.cy.js]: One more change that I should've picked up on in my previous review.

Can we make this a helper that we import? Otherwise we will need to duplicate it in every component testing file that has emits.

**Review by JManion32 — CHANGES_REQUESTED** [review]

**Comment by JManion32** [comment]
Please look into the failing `Cypress (Feature)` test. It is failing due to a `bar-banner` data-testid.

File: site/cypress/component/StatusBanner.cy.js (hunk 7 above)
Comment (jndlansh) [reply site/cypress/component/StatusBanner.cy.js]: Yes, I have a question. This helper function will then reside in a new file - something like `site\cypress\support\component_test_utils.js`. Is that fine?

Comment (JManion32) [reply site/cypress/component/StatusBanner.cy.js]: Yes, either that or adding to `utils.js`, which currently has only E2E helpers. I think the new file is best since component and E2E shouldn't have much or any overlap.

Comment (jndlansh) [reply site/cypress/component/StatusBanner.cy.js]: Alright, I have added it in a new file, i hope this will work just fine

**Review by dagemcn — COMMENTED** [review]
While I can't comment much on the design of the code and I'm sure Justin will have more insights on that front, I can verify that the status banners do work as intended. The component test is passing as well. I'll do a more in depth functional test once this PR is finalized. Looking good so far.

File: site/cypress/component/StatusBanner.cy.js [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+        });
+    });
+
+    describe('emit lifecycle', () => {
+        it('emits the event exactly once per mount', () => {
+            mountWithEmitSpy(StatusBanner, 'colorChange', defaultProps, 'colorChangeHandler');
+            cy.get('@colorChangeHandler').should('have.callCount', 1);
+        });
+
+        it('emits a separate event per mount instance', () => {
```
Comment (JManion32) [inline site/cypress/component/StatusBanner.cy.js]: We can remove this as well. It is a guarantee by Cypress and Vue.

**Review by JManion32 — CHANGES_REQUESTED** [review]
This looks good to me. I found one more component test that looks unnecessary. Other than that, great work!

File: site/vue/src/main.ts [hunk 9]
```diff
@@ -17,10 +18,34 @@ const exports = {
                     throw new Error(`Module ${path} not found`);
                 }
                 const mod = await modules[path]();
+
+                if (Object.keys(events).length > 0) {
+                    const camelize = (str: string) => str.replace(/-(\w)/g, (_, c: string) => (c ? c.toUpperCase() : ''));
+                    const capitalize = (str: string) => str.charAt(0).toUpperCase() + str.slice(1);
```
Comment (williamjallen) [inline site/vue/src/main.ts:24]: Why not just require users to provide the camel case name to begin with?

**Review by williamjallen — CHANGES_REQUESTED** [review]
A few more comments, and then this should be good to go.

File: site/vue/src/main.ts (hunk 9 above)
Comment (jndlansh) [reply site/vue/src/main.ts:24]: kebab-case is the standard Vue emit name and would look natural and consistent to write it like that in Twig, I guess. What is your opinion on that?

Comment (williamjallen) [reply site/vue/src/main.ts:24]: Eh, sure, it doesn't really harm anything to support that.  Let's leave it as-is then.

**Review by williamjallen — APPROVED** [review]
Code looks good to me.  I'd like to get a final approval from @JManion32 before merge though.

**Review by JManion32 — APPROVED** [review]
Code and functionality looks good to me. With the new infrastructure for capturing emits, components no longer depend on Submitty-specific state and we can avoid another refactor. The component tests adequately cover the new component, and I am happy about the `component_test_utils.js` file we added, which should help streamline the creation of future test files.

Great work Ansh!

**Review by dagemcn — APPROVED** [review]
I retested this after the changes were made and the behavior is the same as main, working in both light mode and dark mode. I tested the behavior when resizing the screen and in a chromium based browser and in Firefox, all working as intended.
